core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java

```diff
@@ -276,6 +289,8 @@ private boolean upgrade(ChannelHandlerContext ctx, HttpRequest request) {
         // Prepare and send the upgrade response. Wait for this write to complete before upgrading,
         // since we need the old codec in-place to properly encode the response.
         if (!upgradeCodec.prepareUpgradeResponse(ctx, request)) {
+            ctx.writeAndFlush(INVALID_SETTINGS_HEADER_RESPONSE.retain()).addListener(CLOSE_ON_FAILURE);
```

Comment:
As an invalid/bad request is received, we may close the connection after the write?
```suggestion
            ctx.writeAndFlush(INVALID_SETTINGS_HEADER_RESPONSE.retain()).addListener(CLOSE);
```